File: app/src/main/java/com/kickstarter/viewmodels/ThanksViewModel.java
package com.kickstarter.viewmodels;

import android.util.Pair;

import com.kickstarter.libs.ActivityViewModel;
import com.kickstarter.libs.CurrentUserType;
import com.kickstarter.libs.Environment;
import com.kickstarter.libs.ExperimentsClientType;
import com.kickstarter.libs.OptimizelyEvent;
import com.kickstarter.libs.RefTag;
import com.kickstarter.libs.preferences.BooleanPreferenceType;
import com.kickstarter.libs.utils.BooleanUtils;
import com.kickstarter.libs.utils.ExperimentData;
import com.kickstarter.libs.utils.ExperimentRevenueData;
import com.kickstarter.libs.utils.ListUtils;
import com.kickstarter.libs.utils.ObjectUtils;
import com.kickstarter.libs.utils.UserUtils;
import com.kickstarter.models.Category;
import com.kickstarter.models.Project;
import com.kickstarter.models.User;
import com.kickstarter.services.ApiClientType;
import com.kickstarter.services.DiscoveryParams;
import com.kickstarter.services.apiresponses.DiscoverEnvelope;
import com.kickstarter.ui.IntentKey;
import com.kickstarter.ui.activities.ThanksActivity;
import com.kickstarter.ui.adapters.ThanksAdapter;
import com.kickstarter.ui.adapters.data.ThanksData;
import com.kickstarter.ui.data.CheckoutData;
import com.kickstarter.ui.data.PledgeData;
import com.kickstarter.ui.viewholders.ProjectCardViewHolder;
import com.kickstarter.ui.viewholders.ThanksCategoryViewHolder;

import java.util.List;

import androidx.annotation.NonNull;
import rx.Observable;
import rx.subjects.BehaviorSubject;
import rx.subjects.PublishSubject;

import static com.kickstarter.libs.rx.transformers.Transformers.combineLatestPair;
import static com.kickstarter.libs.rx.transformers.Transformers.ignoreValues;
import static com.kickstarter.libs.rx.transformers.Transformers.neverError;
import static com.kickstarter.libs.rx.transformers.Transformers.takeWhen;
import static com.kickstarter.libs.utils.BooleanUtils.isTrue;

public interface ThanksViewModel {

  interface Inputs extends ProjectCardViewHolder.Delegate, ThanksCategoryViewHolder.Delegate,
    ThanksAdapter.Delegate {
    /** Call when the user clicks the close button. */
    void closeButtonClicked();

    /** Call when the user accepts the prompt to signup to the Games newsletter. */
    void signupToGamesNewsletterClick();
  }

  interface Outputs {
    /** Emits the data to configure the adapter with. */
    Observable<ThanksData> adapterData();

    /** Emits when we should finish the {@link com.kickstarter.ui.activities.ThanksActivity}. */
    Observable<Void> finish();

    /** Show a dialog confirming the user will be signed up to the games newsletter. Required for German users. */
    Observable<Void> showConfirmGamesNewsletterDialog();

    /** Show a dialog prompting the user to sign-up to the games newsletter. */
    Observable<Void> showGamesNewsletterDialog();

    /** Show a dialog prompting the user to rate the app. */
    Observable<Void> showRatingDialog();

    /** Emits when we should start the {@link com.kickstarter.ui.activities.DiscoveryActivity}. */
    Observable<DiscoveryParams> startDiscoveryActivity();

    /** Emits when we should start the {@link com.kickstarter.ui.activities.ProjectActivity}. */
    Observable<Pair<Project, RefTag>> startProjectActivity();
  }

  final class ViewModel extends ActivityViewModel<ThanksActivity> implements Inputs, Outputs {
    private final ApiClientType apiClient;
    private final BooleanPreferenceType hasSeenAppRatingPreference;
    private final BooleanPreferenceType hasSeenGamesNewsletterPreference;
    private final CurrentUserType currentUser;
    private final ExperimentsClientType optimizely;

    public ViewModel(final @NonNull Environment environment) {
      super(environment);

      this.apiClient = environment.apiClient();
      this.currentUser = environment.currentUser();
      this.hasSeenAppRatingPreference = environment.hasSeenAppRatingPreference();
      this.hasSeenGamesNewsletterPreference = environment.hasSeenGamesNewsletterPreference();
      this.optimizely = environment.optimizely();

      final Observable<Project> project = intent()
        .map(i -> i.getParcelableExtra(IntentKey.PROJECT))
        .ofType(Project.class)
        .take(1)
        .compose(bindToLifecycle());

      final Observable<Category> rootCategory = project.flatMap(p -> rootCategory(p, this.apiClient));

      final Observable<Boolean> isGamesCategory = rootCategory
        .map(c -> "games".equals(c.slug()));

      final Observable<Boolean> hasSeenGamesNewsletterDialog = Observable.just(this.hasSeenGamesNewsletterPreference.get());

      final Observable<Boolean> isSignedUpToGamesNewsletter = this.currentUser.observable()
        .map(u -> u != null && isTrue(u.gamesNewsletter()));

      final Observable<Boolean> showGamesNewsletter = Observable.combineLatest(
        isGamesCategory, hasSeenGamesNewsletterDialog, isSignedUpToGamesNewsletter,
        (isGames, hasSeen, isSignedUp) -> isGames && !hasSeen && !isSignedUp
      )
        .take(1);

      this.categoryCardViewHolderClicked
        .map(c -> DiscoveryParams.builder().category(c).build())
        .compose(bindToLifecycle())
        .subscribe(this.startDiscoveryActivity::onNext);

      this.closeButtonClicked
        .compose(bindToLifecycle())
        .subscribe(this.finish);

      this.projectCardViewHolderClicked
        .compose(bindToLifecycle())
        .subscribe(p -> this.startProjectActivity.onNext(Pair.create(p, RefTag.thanks())));

      Observable.combineLatest(
        project,
        rootCategory,
        project.flatMap(p -> this.relatedProjects(p, this.apiClient)),
        ThanksData::new
      )
        .compose(bindToLifecycle())
        .subscribe(this.adapterData::onNext);

      Observable.just(this.hasSeenAppRatingPreference.get())
        .take(1)
        .compose(combineLatestPair(showGamesNewsletter))
        .filter(ag -> !ag.first && !ag.second)
        .compose(ignoreValues())
        .compose(bindToLifecycle())
        .subscribe(__ -> this.showRatingDialog.onNext(null));

      showGamesNewsletter
        .filter(x -> x)
        .compose(bindToLifecycle())
        .subscribe(__ -> this.showGamesNewsletterDialog.onNext(null));

      this.showGamesNewsletterDialog
        .compose(bindToLifecycle())
        .subscribe(__ -> this.hasSeenGamesNewsletterPreference.set(true));

      this.currentUser.observable()
        .filter(ObjectUtils::isNotNull)
        .compose(takeWhen(this.signupToGamesNewsletterClick))
        .flatMap(u -> this.signupToGamesNewsletter(u, this.apiClient))
        .compose(bindToLifecycle())
        .subscribe(this.signedUpToGamesNewsletter::onNext);

      this.currentUser.observable()
        .filter(ObjectUtils::isNotNull)
        .compose(takeWhen(this.signedUpToGamesNewsletter))
        .filter(UserUtils::isLocationGermany)
        .compose(bindToLifecycle())
        .subscribe(__ -> this.showConfirmGamesNewsletterDialog.onNext(null));

      // Event tracking
      this.categoryCardViewHolderClicked
        .compose(bindToLifecycle())
        .subscribe(__ -> this.koala.trackCheckoutFinishJumpToDiscovery());

      this.projectCardViewHolderClicked
        .compose(bindToLifecycle())
        .subscribe(this.koala::trackCheckoutFinishJumpToProject);

      this.signedUpToGamesNewsletter
        .compose(bindToLifecycle())
        .subscribe(__ -> this.koala.trackNewsletterToggle(true));

      final Observable<CheckoutData> checkoutData = intent()
        .map(i -> i.getParcelableExtra(IntentKey.CHECKOUT_DATA))
        .ofType(CheckoutData.class)
        .take(1);

      final Observable<PledgeData> pledgeData = intent()
        .map(i -> i.getParcelableExtra(IntentKey.PLEDGE_DATA))
        .ofType(PledgeData.class)
        .take(1);

      final Observable<Pair<CheckoutData, PledgeData>> checkoutAndPledgeData =
        Observable.combineLatest(checkoutData, pledgeData, Pair::create);

      checkoutAndPledgeData
        .compose(bindToLifecycle())
        .subscribe(checkoutDataPledgeData -> this.lake.trackThanksPageViewed(checkoutDataPledgeData.first, checkoutDataPledgeData.second));

      checkoutAndPledgeData
        .compose(combineLatestPair(this.currentUser.observable()))
        .map(this::experimentRevenueData)
        .take(1)
        .compose(bindToLifecycle())
        .subscribe(data -> this.optimizely.trackRevenue(OptimizelyEvent.APP_COMPLETED_CHECKOUT, data));
    }

    private ExperimentRevenueData experimentRevenueData(final @NonNull Pair<Pair<CheckoutData, PledgeData>, User> dataAndUser) {
      final User currentUser = dataAndUser.second;
      final PledgeData pledgeData = dataAndUser.first.second;
      final RefTag intentRefTag = pledgeData.projectData().refTagFromIntent();
      final RefTag cookieRefTag = pledgeData.projectData().refTagFromCookie();
      final ExperimentData experimentData = new ExperimentData(currentUser, intentRefTag, cookieRefTag);
      final CheckoutData checkoutData = dataAndUser.first.first;
      return new ExperimentRevenueData(experimentData, checkoutData, pledgeData);
    }

    /**
     * Given a project, returns an observable that emits the project's root category.
     */
    private static @NonNull Observable<Category> rootCategory(final @NonNull Project project, final @NonNull ApiClientType client) {
      final Category category = project.category();

      if (category == null) {
        return Observable.empty();
      }

      if (category.parent() != null) {
        return Observable.just(category.parent());
      }

      return client.fetchCategory(String.valueOf(category.rootId()))
        .compose(neverError());
    }

    /**
     * Returns a shuffled list of 3 recommended projects, with fallbacks to similar and staff picked projects
     * for users with fewer than 3 recommendations.
     */
    private @NonNull Observable<List<Project>> relatedProjects(final @NonNull Project project, final @NonNull ApiClientType client) {
      final DiscoveryParams recommendedParams = DiscoveryParams.builder()
        .backed(-1)
        .recommended(true)
        .perPage(6)
        .build();

      final DiscoveryParams similarToParams = DiscoveryParams.builder()
        .backed(-1)
        .similarTo(project)
        .perPage(3)
        .build();

      final Category category = project.category();
      final DiscoveryParams staffPickParams = DiscoveryParams.builder()
        .category(category == null ? null : category.root())
        .backed(-1)
        .staffPicks(true)
        .perPage(3)
        .build();

      final Observable<Project> recommendedProjects = client.fetchProjects(recommendedParams)
        .retry(2)
        .map(DiscoverEnvelope::projects)
        .map(ListUtils::shuffle)
        .flatMap(Observable::from)
        .take(3);

      final Observable<Project> similarToProjects = client.fetchProjects(similarToParams)
        .retry(2)
        .map(DiscoverEnvelope::projects)
        .flatMap(Observable::from);

      final Observable<Project> staffPickProjects = client.fetchProjects(staffPickParams)
        .retry(2)
        .map(DiscoverEnvelope::projects)
        .flatMap(Observable::from);

      return Observable.concat(recommendedProjects, similarToProjects, staffPickProjects)
        .compose(neverError())
        .distinct()
        .take(3)
        .toList();
    }

    private Observable<User> signupToGamesNewsletter(final @NonNull User user, final @NonNull ApiClientType client) {
      return client
        .updateUserSettings(user.toBuilder().gamesNewsletter(true).build())
        .compose(neverError());
    }

    private final PublishSubject<Category> categoryCardViewHolderClicked = PublishSubject.create();
    private final PublishSubject<Void> closeButtonClicked = PublishSubject.create();
    private final PublishSubject<Project> projectCardViewHolderClicked  = PublishSubject.create();
    private final PublishSubject<Void> signupToGamesNewsletterClick = PublishSubject.create();

    private final BehaviorSubject<ThanksData> adapterData = BehaviorSubject.create();
    private final PublishSubject<Void> finish = PublishSubject.create();
    private final PublishSubject<Void> showConfirmGamesNewsletterDialog = PublishSubject.create();
    private final PublishSubject<Void> showGamesNewsletterDialog = PublishSubject.create();
    private final PublishSubject<Void> showRatingDialog = PublishSubject.create();
    private final PublishSubject<User> signedUpToGamesNewsletter = PublishSubject.create();
    private final PublishSubject<DiscoveryParams> startDiscoveryActivity = PublishSubject.create();
    private final PublishSubject<Pair<Project, RefTag>> startProjectActivity = PublishSubject.create();

    public final Inputs inputs = this;
    public final Outputs outputs = this;

    @Override public void categoryViewHolderClicked(final @NonNull Category category) {
      this.categoryCardViewHolderClicked.onNext(category);
    }
    @Override public void closeButtonClicked() {
      this.closeButtonClicked.onNext(null);
    }
    @Override public void signupToGamesNewsletterClick() {
      this.signupToGamesNewsletterClick.onNext(null);
    }
    @Override public void projectCardViewHolderClicked(final @NonNull Project project) {
      this.projectCardViewHolderClicked.onNext(project);
    }

    @Override public @NonNull Observable<ThanksData> adapterData() {
      return this.adapterData;
    }
    @Override public @NonNull Observable<Void> finish() {
      return this.finish;
    }
    @Override public @NonNull Observable<Void> showConfirmGamesNewsletterDialog() {
      return this.showConfirmGamesNewsletterDialog;
    }
    @Override public @NonNull Observable<Void> showGamesNewsletterDialog() {
      return this.showGamesNewsletterDialog;
    }
    @Override public @NonNull Observable<Void> showRatingDialog() {
      return this.showRatingDialog;
    }
    @Override public @NonNull Observable<DiscoveryParams> startDiscoveryActivity() {
      return this.startDiscoveryActivity;
    }
    @Override public @NonNull Observable<Pair<Project, RefTag>> startProjectActivity() {
      return this.startProjectActivity;
    }
  }
}

File: app/src/main/java/com/kickstarter/viewmodels/ThanksViewModel.kt
package com.kickstarter.viewmodels

import android.content.Intent
import android.os.Parcelable
import android.util.Pair
import androidx.annotation.VisibleForTesting
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.kickstarter.libs.Environment
import com.kickstarter.libs.RefTag
import com.kickstarter.libs.featureflag.FlagKey
import com.kickstarter.libs.rx.transformers.Transformers
import com.kickstarter.libs.utils.EventContextValues.ContextPageName.THANKS
import com.kickstarter.libs.utils.ListUtils
import com.kickstarter.libs.utils.RefTagUtils
import com.kickstarter.libs.utils.ThirdPartyEventValues
import com.kickstarter.libs.utils.extensions.addToDisposable
import com.kickstarter.libs.utils.extensions.combineProjectsAndParams
import com.kickstarter.libs.utils.extensions.isNotNull
import com.kickstarter.libs.utils.extensions.isTrue
import com.kickstarter.libs.utils.extensions.updateStartedProjectAndDiscoveryParamsList
import com.kickstarter.models.Category
import com.kickstarter.models.Project
import com.kickstarter.models.User
import com.kickstarter.models.extensions.isLocationGermany
import com.kickstarter.services.ApiClientTypeV2
import com.kickstarter.services.ApolloClientTypeV2
import com.kickstarter.services.DiscoveryParams
import com.kickstarter.ui.IntentKey
import com.kickstarter.ui.adapters.ThanksAdapter
import com.kickstarter.ui.adapters.data.ThanksData
import com.kickstarter.ui.data.CheckoutData
import com.kickstarter.ui.data.PledgeData
import com.kickstarter.ui.data.ProjectData
import com.kickstarter.ui.viewholders.ProjectCardViewHolder
import com.kickstarter.ui.viewholders.ThanksCategoryViewHolder
import com.kickstarter.viewmodels.usecases.SendThirdPartyEventUseCaseV2
import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.subjects.BehaviorSubject
import io.reactivex.subjects.PublishSubject

interface ThanksViewModel {
    interface Inputs :
        ProjectCardViewHolder.Delegate,
        ThanksCategoryViewHolder.Delegate,
        ThanksAdapter.Delegate {
        /** Call when the user clicks the close button.  */
        fun closeButtonClicked()

        /** Call when the user accepts the prompt to signup to the Games newsletter.  */
        fun signupToGamesNewsletterClick()
    }

    interface Outputs {
        /** Emits the data to configure the adapter with.  */
        fun adapterData(): Observable<ThanksData>

        /** Emits when we should finish the [com.kickstarter.ui.activities.ThanksActivity].  */
        fun finish(): Observable<Unit>

        /** Show a dialog confirming the user will be signed up to the games newsletter. Required for German users.  */
        fun showConfirmGamesNewsletterDialog(): Observable<Unit>

        /** Show a dialog prompting the user to sign-up to the games newsletter.  */
        fun showGamesNewsletterDialog(): Observable<Unit>

        /** Show a dialog prompting the user to rate the app.  */
        fun showRatingDialog(): Observable<Unit>

        /** Emits when we should start the [com.kickstarter.ui.activities.DiscoveryActivity].  */
        fun startDiscoveryActivity(): Observable<DiscoveryParams>

        /** Emits when we should start the [com.kickstarter.ui.activities.ProjectActivity].  */
        fun startProjectActivity(): Observable<Pair<Project, RefTag>>

        /** Emits when the success prompt for saving should be displayed.  */
        fun showSavedPrompt(): Observable<Unit>
    }

    class ThanksViewModel(environment: Environment, private val intent: Intent) :
        ViewModel(),
        Inputs,
        Outputs {
        private val apiClient = requireNotNull(environment.apiClientV2())
        private val apolloClient = requireNotNull(environment.apolloClientV2())
        private val hasSeenAppRatingPreference = environment.hasSeenAppRatingPreference()
        private val hasSeenGamesNewsletterPreference = environment.hasSeenGamesNewsletterPreference()
        private val currentUser = requireNotNull(environment.currentUserV2())
        private val sharedPreferences = requireNotNull(environment.sharedPreferences())
        private val cookieManager = requireNotNull(environment.cookieManager())
        private val ffClient = requireNotNull(environment.featureFlagClient())

        private val categoryCardViewHolderClicked = PublishSubject.create<Category>()
        private val closeButtonClicked = PublishSubject.create<Unit>()
        private val projectCardViewHolderClicked = PublishSubject.create<Project>()
        private val signupToGamesNewsletterClick = PublishSubject.create<Unit>()
        private val adapterData = BehaviorSubject.create<ThanksData>()
        private val finish = PublishSubject.create<Unit>()
        private val showConfirmGamesNewsletterDialog = PublishSubject.create<Unit>()
        private val showGamesNewsletterDialog = BehaviorSubject.create<Unit>()
        private val showRatingDialog = BehaviorSubject.create<Unit>()
        private val signedUpToGamesNewsletter = PublishSubject.create<User>()
        private val startDiscoveryActivity = PublishSubject.create<DiscoveryParams>()
        private val startProjectActivity = PublishSubject.create<Pair<Project, RefTag>>()
        private val onHeartButtonClicked = PublishSubject.create<Project>()
        private val showSavedPrompt = PublishSubject.create<Unit>()
        private val analyticEvents = environment.analytics()

        @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
        val onThirdPartyEventSent = BehaviorSubject.create<Boolean?>()
        @JvmField
        val inputs: Inputs = this

        @JvmField
        val outputs: Outputs = this

        private fun intent() = intent.let { Observable.just(it) }

        private val disposables = CompositeDisposable()

        init {
            val project = intent()
                .filter { it.getParcelableExtra<Parcelable?>(IntentKey.PROJECT).isNotNull() }
                .map<Project?> { it.getParcelableExtra(IntentKey.PROJECT) }
                .ofType(Project::class.java)
                .take(1)

            val rootCategory = project
                .switchMap {
                    rootCategory(it, apolloClient)
                }
                .compose(Transformers.neverErrorV2())
                .filter {
                    it.isNotNull()
                }.map { it }

            val isGamesCategory = rootCategory
                .map { "games" == it.slug() }

            val hasSeenGamesNewsletterDialog = Observable.just(
                hasSeenGamesNewsletterPreference?.get() ?: false
            ).filter { it.isNotNull() }
                .map { it }

            val isSignedUpToGamesNewsletter = currentUser.observable()
                .map { it.getValue() != null && it.getValue()?.gamesNewsletter().isTrue() }

            val showGamesNewsletter = Observable.combineLatest(
                isGamesCategory,
                hasSeenGamesNewsletterDialog,
                isSignedUpToGamesNewsletter
            ) { isGames, hasSeen, isSignedUp ->
                isGames && !hasSeen && !isSignedUp
            }

            categoryCardViewHolderClicked
                .map { DiscoveryParams.builder().category(it).build() }
                .subscribe { startDiscoveryActivity.onNext(it) }
                .addToDisposable(disposables)

            closeButtonClicked
                .subscribe(finish)

            projectCardViewHolderClicked
                .subscribe {
                    startProjectActivity.onNext(
                        Pair(
                            it,
                            RefTag.thanks()
                        )
                    )
                }
                .addToDisposable(disposables)

            val projectOnUserChangeSave = this.onHeartButtonClicked
                .switchMap {
                    this.toggleProjectSave(it)
                }
                .share()

            val recommendedProjects = project.switchMap {
                relatedProjects(
                    it,
                    apiClient
                )?.toObservable()
            }.filter { it.isNotNull() }
                .map {
                    combineProjectsAndParams(
                        it,
                        DiscoveryParams.builder().build()
                    )
                }

            projectOnUserChangeSave
                .subscribe { this.analyticEvents?.trackWatchProjectCTA(it, THANKS) }
                .addToDisposable(disposables)

            projectOnUserChangeSave
                .filter { p -> p.isStarred() && p.isLive && !p.isApproachingDeadline }
                .compose(Transformers.ignoreValuesV2())
                .subscribe(this.showSavedPrompt)

            Observable.just(hasSeenAppRatingPreference?.get())
                .filter { it.isNotNull() }
                .map { it }
                .take(1)
                .compose(Transformers.combineLatestPair(showGamesNewsletter))
                .filter { !it.first && !it.second }
                .filter { environment.featureFlagClient()?.getBoolean(FlagKey.ANDROID_HIDE_APP_RATING_DIALOG) == false }
                .compose(Transformers.ignoreValuesV2())
                .subscribe { showRatingDialog.onNext(Unit) }
                .addToDisposable(disposables)

            showGamesNewsletter
                .filter { it }
                .subscribe { showGamesNewsletterDialog.onNext(Unit) }
                .addToDisposable(disposables)

            showGamesNewsletterDialog
                .subscribe { hasSeenGamesNewsletterPreference?.set(true) }
                .addToDisposable(disposables)

            currentUser.observable()
                .filter { it.getValue().isNotNull() }
                .map { it.getValue() }
                .compose(Transformers.takeWhenV2(signupToGamesNewsletterClick))
                .flatMap { signupToGamesNewsletter(it, apiClient) }
                .subscribe { signedUpToGamesNewsletter.onNext(it) }
                .addToDisposable(disposables)

            currentUser.observable()
                .filter { it.getValue().isNotNull() }
                .map { it.getValue() }
                .compose(Transformers.takeWhenV2(signedUpToGamesNewsletter))
                .filter {
                    it.isLocationGermany()
                }
                .subscribe { showConfirmGamesNewsletterDialog.onNext(Unit) }
                .addToDisposable(disposables)

            val checkoutData = intent()
                .filter { it.getParcelableExtra<Parcelable?>(IntentKey.CHECKOUT_DATA).isNotNull() }
                .map<CheckoutData?> { it.getParcelableExtra(IntentKey.CHECKOUT_DATA) }
                .ofType(CheckoutData::class.java)
                .take(1)

            val pledgeData = intent()
                .filter { it.getParcelableExtra<Parcelable?>(IntentKey.PLEDGE_DATA).isNotNull() }
                .map<PledgeData?> { it.getParcelableExtra(IntentKey.PLEDGE_DATA) }
                .ofType(PledgeData::class.java)
                .take(1)

            val checkoutAndPledgeData =
                Observable.combineLatest<CheckoutData, PledgeData, Pair<CheckoutData, PledgeData>>(
                    checkoutData,
                    pledgeData
                ) { a, b -> Pair(a, b) }

            checkoutAndPledgeData
                .subscribe { checkoutDataPledgeData: Pair<CheckoutData, PledgeData> ->
                    analyticEvents?.trackThanksScreenViewed(
                        checkoutDataPledgeData.first,
                        checkoutDataPledgeData.second
                    )
                }
                .addToDisposable(disposables)

            Observable.combineLatest(
                project,
                rootCategory,
                recommendedProjects.startWith(listOf()),
                checkoutData,
            ) { backedProject, category, projects, checkoutData ->
                ThanksData(backedProject, checkoutData, category, projects)
            }.subscribe { adapterData.onNext(it) }
                .addToDisposable(disposables)

            adapterData
                .compose(Transformers.takePairWhenV2(projectOnUserChangeSave))
                .map { Pair(it.first, it.second.updateStartedProjectAndDiscoveryParamsList(it.first.recommendedProjects)) }
                .map {
                    ThanksData(it.first.backedProject, it.first.checkoutData, it.first.category, it.second)
                }.distinctUntilChanged()
                .subscribe {
                    adapterData.onNext(it)
                }
                .addToDisposable(disposables)

            SendThirdPartyEventUseCaseV2(sharedPreferences, ffClient)
                .sendThirdPartyEvent(
                    project = project,
                    apolloClient = apolloClient,
                    currentUser = currentUser,
                    eventName = ThirdPartyEventValues.EventName.PURCHASE,
                    checkoutAndPledgeData = checkoutAndPledgeData,
                )
                .compose(Transformers.neverErrorV2())
                .subscribe {
                    onThirdPartyEventSent.onNext(it.first)
                }
                .addToDisposable(disposables)

            checkoutAndPledgeData
                .compose(Transformers.takePairWhenV2(projectCardViewHolderClicked))
                .subscribe { dataCheckoutProjectPair: Pair<Pair<CheckoutData, PledgeData>, Project> ->

                    val cookieRefTag = RefTagUtils.storedCookieRefTagForProject(
                        dataCheckoutProjectPair.second,
                        cookieManager,
                        sharedPreferences
                    )

                    val projectData = ProjectData.builder()
                        .refTagFromIntent(RefTag.thanks())
                        .refTagFromCookie(cookieRefTag)
                        .project(dataCheckoutProjectPair.second)
                        .build()

                    analyticEvents?.trackThanksActivityProjectCardClicked(
                        projectData,
                        dataCheckoutProjectPair.first.first,
                        dataCheckoutProjectPair.first.second
                    )
                }.addToDisposable(disposables)
        }

        /**
         * Returns a shuffled list of 3 recommended projects, with fallbacks to similar and staff picked projects
         * for users with fewer than 3 recommendations.
         */
        private fun relatedProjects(
            project: Project,
            client: ApiClientTypeV2
        ): Single<MutableList<Project>>? {
            val recommendedParams = DiscoveryParams.builder()
                .backed(-1)
                .recommended(true)
                .perPage(6)
                .build()

            val similarToParams = DiscoveryParams.builder()
                .backed(-1)
                .similarTo(project)
                .perPage(3)
                .build()

            val category = project.category()

            val staffPickParams = DiscoveryParams.builder()
                .category(category?.root())
                .backed(-1)
                .staffPicks(true)
                .perPage(3)
                .build()

            val recommendedProjects = client.fetchProjects(recommendedParams)
                .retry(2)
                .map { it.projects() }
                .map { ListUtils.shuffle(it) }
                .flatMap { Observable.fromIterable(it) }
                .take(3)
                .filter { it.isNotNull() }

            val similarToProjects = client.fetchProjects(similarToParams)
                .retry(2)
                .map { it.projects() }
                .flatMap { Observable.fromIterable(it) }

            val staffPickProjects = client.fetchProjects(staffPickParams)
                .retry(2)
                .map { it.projects() }
                .flatMap { Observable.fromIterable(it) }

            return Observable.concat(recommendedProjects, similarToProjects, staffPickProjects)
                .compose(Transformers.neverErrorV2())
                .distinct()
                .take(3)
                .toList()
        }

        private fun signupToGamesNewsletter(user: User, client: ApiClientTypeV2): Observable<User> {
            return client
                .updateUserSettings(user.toBuilder().gamesNewsletter(true).build())
                .compose(Transformers.neverErrorV2())
        }
        override fun categoryViewHolderClicked(category: Category) = categoryCardViewHolderClicked.onNext(category)
        override fun closeButtonClicked() = closeButtonClicked.onNext(Unit)
        override fun signupToGamesNewsletterClick() = signupToGamesNewsletterClick.onNext(Unit)
        override fun onHeartButtonClicked(project: Project) = onHeartButtonClicked.onNext(project)
        override fun projectCardViewHolderClicked(project: Project) = projectCardViewHolderClicked.onNext(project)

        override fun adapterData(): Observable<ThanksData> = this.adapterData
        override fun finish(): Observable<Unit> = this.finish
        override fun showConfirmGamesNewsletterDialog(): Observable<Unit> = this.showConfirmGamesNewsletterDialog
        override fun showGamesNewsletterDialog(): Observable<Unit> = this.showGamesNewsletterDialog
        override fun showRatingDialog(): Observable<Unit> = this.showRatingDialog
        override fun startDiscoveryActivity(): Observable<DiscoveryParams> = this.startDiscoveryActivity
        override fun startProjectActivity(): Observable<Pair<Project, RefTag>> = this.startProjectActivity
        override fun showSavedPrompt(): Observable<Unit> = this.showSavedPrompt

        private fun saveProject(project: Project): Observable<Project> {
            return this.apolloClient.watchProject(project)
                .compose(Transformers.neverErrorV2())
        }

        private fun unSaveProject(project: Project): Observable<Project> {
            return this.apolloClient.unWatchProject(project).compose(Transformers.neverErrorV2())
        }

        private fun toggleProjectSave(project: Project): Observable<Project> {
            return if (project.isStarred()) {
                unSaveProject(project)
            } else {
                saveProject(project)
            }
        }

        companion object {
            /**
             * Given a project, returns an observable that emits the project's root category.
             */
            private fun rootCategory(
                project: Project,
                client: ApolloClientTypeV2
            ): Observable<Category?> {
                val category = project.category() ?: return Observable.empty()

                return when {
                    category.parent() != null -> {
                        Observable.just(category.parent())
                    }
                    category.isRoot -> {
                        Observable.just(category)
                    }
                    else -> {
                        client.fetchCategory(category.rootId().toString())
                    }
                }
            }
        }

        override fun onCleared() {
            apolloClient.cleanDisposables()
            disposables.clear()
            super.onCleared()
        }
    }

    class Factory(private val environment: Environment, private val intent: Intent) : ViewModelProvider.Factory {
        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            return ThanksViewModel(environment, intent) as T
        }
    }
}
